cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,320 @@
+      String originalBrokerBrokerSetId = _brokerSetResolutionHelper.brokerSetId(replica.originalBroker());
+
+      // If the re-balancing action is performed within the broker set boundary, we are good
+      if (broker.isAlive() && originalBrokerBrokerSetId.equals(currentBrokerSetId)) {
+        continue;
+      }
+      // If the brokerSet awareness condition is violated. Move replica to an eligible broker
+      Set<Broker> eligibleBrokers = _brokersByBrokerSet.get(originalBrokerBrokerSetId)
+                                                       .stream()
+                                                       .filter(b -> clusterModel.aliveBrokers().contains(b))
+                                                       .collect(Collectors.toSet());
+      if (maybeApplyBalancingAction(clusterModel, replica, eligibleBrokers, ActionType.INTER_BROKER_REPLICA_MOVEMENT, optimizedGoals,
+                                    optimizationOptions) == null) {
+        // If balancing action can not be applied, provide recommendation to add new Brokers.
+        ProvisionRecommendation recommendation =
+            new ProvisionRecommendation.Builder(ProvisionStatus.UNDER_PROVISIONED).numBrokers(clusterModel.maxReplicationFactor()).build();
+        throw new OptimizationFailureException(
+            String.format("[%s] Cannot move replica %s to %s on brokerSet %s", name(), replica, eligibleBrokers, originalBrokerBrokerSetId),
+            recommendation);
+      }
+    }
+  }
+
+  /**
+   * Check whether the given action is acceptable by this goal. The following actions are acceptable:
+   * <ul>
+   *   <li>All leadership moves</li>
+   *   <li>Replica moves that do not violate {@link #doesReplicaMoveViolateActionAcceptance(Replica, Broker)}</li>
+   *   <li>Swaps that do not violate {@link #doesReplicaMoveViolateActionAcceptance(Replica, Broker)}
+   *   in both direction</li>
+   * </ul>
+   *
+   * @param action       Action to be checked for acceptance.
+   * @param clusterModel The state of the cluster.
+   * @return {@link ActionAcceptance#ACCEPT} if the action is acceptable by this goal,
+   * {@link ActionAcceptance#BROKER_REJECT} if the action is rejected due to violating brokerSet awareness in the destination
+   * broker after moving source replica to destination broker, {@link ActionAcceptance#REPLICA_REJECT} otherwise.
+   */
+  @Override
+  public ActionAcceptance actionAcceptance(BalancingAction action, ClusterModel clusterModel) {
+    switch (action.balancingAction()) {
+      case LEADERSHIP_MOVEMENT:
+        return ACCEPT;
+      case INTER_BROKER_REPLICA_MOVEMENT:
+      case INTER_BROKER_REPLICA_SWAP:
+        if (doesReplicaMoveViolateActionAcceptance(clusterModel.broker(action.sourceBrokerId()).replica(action.topicPartition()),
+                                                   clusterModel.broker(action.destinationBrokerId()))) {
+          return BROKER_REJECT;
```

Comment:
Reading the java doc for ActionAcceptance, I think this should also be REPLICA_REJECT?

Reply:
I have changed the code slightly but most remains the same. Is the Destination Broker is not the correct brokerSet then it is BROKER_REJECT - otherwise REPLICA_REJECT.

My understanding comes for a similar Goal here -
https://github.com/linkedin/cruise-control/blob/4e5927b48bf2581ab76acbbecbf42b355b871b65/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/AbstractRackAwareGoal.java#L78

Reply:
From the java doc:

"BROKER_REJECT: Action is rejected in broker-level; hence, the destination broker does not accept actions of the same ActionType from the source broker specified in the given action."

After re-read it, I think my original comment was wrong... It seems like it should always be BROKER_REJECT in your case (line263 doesn't need to change, but line278 need to change). BROKER_REJECT means "ANY partition movement from source broker to destination broker is not allowed", which is exactly your case since the source broker and dest broker are from different brokerSet.

@efeg to confirm.

Reply:
The code in the PR seems correct to me.

The reason why the line278 is `REPLICA_REJECT` is because it is for a replica swap action -- i.e. this goal can move the source replica to the destination broker, but it cannot move the destination replica to the source broker; hence, it is a reject for the destination replica's move to the source broker. If a goal can find another destination replica to swap with the source replica it can proceed. For more details, please see how the `INTER_BROKER_REPLICA_SWAP` is used in distribution goals to prune the exploration space of alternative actions to satisfy the goal.